### ISSUE TITLE
[HOLD for payment 2023-05-03] [$500] Workspace - Deleted Workspace offline is not crossed out

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#17656]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Issue found when executing PR** https://github.com/Expensify/App/pull/17618

## Action Performed:
1. Go to staging.new.expensify.com
2. Log in with any account
3. Go to Setting - Workspace
4. Create  a new Workspace
5. Disable the internet 
6. Delete the Workspace

## Expected Result:
Deleted Workspace offline is crossed out

## Actual Result:
Deleted Workspace offline is not crossed out

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.1.3

**Reproducible in staging?:** Yes

**Reproducible in production?:** No

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/233100456-c83129b0-908c-4035-bdd9-28f12cd23cca.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~018ed51b0bbc8584dc</li>
        <li>Upwork Job ID: 1648765595714977792</li>
        <li>Last Price Increase: 2023-04-19</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @strepanier03 (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by kbecciv** [comment]
@strepanier03 Issue is not reproduced in Production, I did not use the Deploy Blocker label since this issue coming from PR listed above.

![IMG_6152](https://user-images.githubusercontent.com/93399543/233100963-9eb564ad-339f-4551-abbb-cb839b7aa49d.png)

**Comment by mountiny** [comment]
Weird indeed @akinwale this seems to also stem from the revert, its probably some combination of another change. Do you want to have a look at this and incorporate it into the PR we have open so the design of the menu item is correct when worskapce is deleted?

**Comment by akinwale** [comment]
> Weird indeed @akinwale this seems to also stem from the revert, its probably some combination of another change. Do you want to have a look at this and incorporate it into the PR we have open so the design of the menu item is correct when worskapce is deleted?

Could you point out the design issue? Your response doesn't seem to have a screenshot attached nor a quoted comment.

**Comment by akinwale** [comment]
> Weird indeed @akinwale this seems to also stem from the revert, its probably some combination of another change. Do you want to have a look at this and incorporate it into the PR we have open so the design of the menu item is correct when worskapce is deleted?

Never mind my previous comment, didn't initially realise this is a separate issue. I'll look into it.

**Comment by akinwale** [comment]
@mountiny This is also related to a change to `MenuItem` in this commit: https://github.com/Expensify/App/commit/a453d1e610a5dc14c62c21f5172dbbe444e3d0dd.

Initially, `titleTextStyle` in the `MenuItem` component was taking the `props.style` as part of the style combinations. `OfflineWithFeedback` updates the style prop of the component with the appropriate style for strikethrough.

https://github.com/Expensify/App/blob/b6f6f6a1bcaddd83d926e81dc1bea57803633314/src/components/OfflineWithFeedback.js#L76

Since the `titleTextStyle` no longer has `props.style` combined with the rest of the applicable styles, strikethrough never gets applied to the text in the menu item.

To fix this, I have to modify the `applyStrikethrough` method in `OfflineWithFeedback` to also check for a `titleStyle` prop. The changes would go like so:
```
let props = {style: StyleUtils.combineStyles(child.props.style, styles.offlineFeedback.deleted, styles.userSelectNone)};
if (child.props.titleStyle) {
    props = {...props, titleStyle: StyleUtils.combineStyles(child.props.titleStyle, styles.offlineFeedback.deleted)};
}
```

Alternatively, I could add a boolean prop to `MenuItem` like `inOfflineState` which will be used to conditionally apply the `styles.offlineFeedback.deleted` style to `titleTextStyle`. Then update the `applyStrikeThrough` method to set the prop to `true` on the child components.

Both approaches would ultimately require modifying `OfflineWithFeedback`. Let me know which is preferred.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~018ed51b0bbc8584dc

**Comment by MelvinBot** [comment]
Current assignee @strepanier03 is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @eVoloshchak is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @mountiny is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $500

**Comment by MelvinBot** [comment]
📣 @akinwale You have been assigned to this job by @mountiny!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by mountiny** [comment]
Thanks, I have exported this with slightly lower price tag since its just a small styling issue. Assigned you and please address it in the PR (I saw you already did it). Thanks @akinwale 

@eVoloshchak Could you please chime in what would be the best way to fix this in your opinion, thank you!

**Comment by strepanier03** [comment]
Thanks for moving this one along @mountiny, I'll continue following along.

**Comment by eVoloshchak** [comment]
@akinwale, i agree with you, this has been caused by https://github.com/Expensify/App/commit/a453d1e610a5dc14c62c21f5172dbbe444e3d0dd. 

I don't agree with the proposed solution, since this should be working as it is. 
[applyStrikeThrough](https://github.com/Expensify/App/blob/1cd304c4d9871caa13362d76f1a8a11a940306b7/src/components/OfflineWithFeedback.js#L71) is iterating over all children recursively, so it should apply strikethrough to Text [here](https://github.com/Expensify/App/blob/1cd304c4d9871caa13362d76f1a8a11a940306b7/src/components/MenuItem.js#L146). It successfully does that on web, but not in native apps. 
I think we need to resolve the root cause, which most likely is in applyStrikeThrough not iterating over all of the children on native apps

**Comment by akinwale** [comment]
> @akinwale, i agree with you, this has been caused by [a453d1e](https://github.com/Expensify/App/commit/a453d1e610a5dc14c62c21f5172dbbe444e3d0dd).
> 
> I don't agree with the proposed solution, since this should be working as it is. [applyStrikeThrough](https://github.com/Expensify/App/blob/1cd304c4d9871caa13362d76f1a8a11a940306b7/src/components/OfflineWithFeedback.js#L71) is iterating over all children recursively, so it should apply strikethrough to Text [here](https://github.com/Expensify/App/blob/1cd304c4d9871caa13362d76f1a8a11a940306b7/src/components/MenuItem.js#L146). It successfully does that on web, but not in native apps. I think we need to resolve the root cause, which most likely is in applyStrikeThrough not iterating over all of the children on native apps

It looks like there is a bug with the `applyStrikeThrough` method. The recursive iteration over the children isn't actually working because `child.props.children` is actually undefined (verified by checking the log output on the value on the `main` branch, without the modifications). Hence, this is why the solution works for now.

On web, when inspecting the HTML elements, the text appears to be inheriting the line through style from the parent div container (the style isn't applied directly on the resulting text node). It seems this inheritance is not happening on native, hence the discrepancy.

I propose that the `applyStrikethrough` bug should be filed as a new issue, while we move forward with this solution to close this out.

**Comment by eVoloshchak** [comment]
Thanks for investigating this
I still think we should explore other options, adding a check for `titleStyle` isn't universal, we'll have to add additional checks each time we have a `style` prop that isn't called `style`. 
Can we resolve this by changing the `MenuItem` itself?

**Comment by akinwale** [comment]
> Thanks for investigating this I still think we should explore other options, adding a check for `titleStyle` isn't universal, we'll have to add additional checks each time we have a `style` prop that isn't called `style`. Can we resolve this by changing the `MenuItem` itself?

I suppose I can check if props.style contains the `styles.offlineFeedback.deleted` value, and then apply it conditionally to the `titleStyle`. With this, I can remove the modification to `OfflineWithFeedback`.

**Comment by eVoloshchak** [comment]
@akinwale, is this issue related to https://github.com/Expensify/App/pull/17639?
If not, I think we should fix it in a separate PR

**Comment by akinwale** [comment]
> @akinwale, is this issue related to #17639? If not, I think we should fix it in a separate PR

I would say it is, considering this comment: https://github.com/Expensify/App/issues/17656#issuecomment-1515237627.

To provide some context, the PR was originally created to address https://github.com/Expensify/App/issues/17580, where I traced the root cause to the `defaultProps` problem. Since https://github.com/Expensify/App/pull/17310 was reverted, the new PR includes the commits from https://github.com/Expensify/App/pull/17310, including the `defaultProps` fix (https://github.com/Expensify/App/issues/17580) and style fixes (for this issue).

**Comment by eVoloshchak** [comment]
Was it caused by https://github.com/Expensify/App/pull/17310?
I can still reproduce this issue even with that PR reverted

**Comment by akinwale** [comment]
> Was it caused by #17310? I can still reproduce this issue even with that PR reverted

No.

This issue (strikethrough styling) is caused by the changes in this commit: https://github.com/Expensify/App/commit/a453d1e610a5dc14c62c21f5172dbbe444e3d0dd. On Web, the title text inherits the parent's `line-through` style, so that looks fine. On native (iOS / Android), the line-through style is not inherited. Since  `child.props.children` is undefined in the `applyStrikeThrough` method in `OfflineWithFeedback`, the style is not applied recursively to the children of the `MenuItem`, hence why there is no strikethrough decoration on the text.

The layout issue (https://github.com/Expensify/App/issues/17580) is caused by https://github.com/Expensify/App/pull/17310 being merged after the commit (https://github.com/Expensify/App/commit/a453d1e610a5dc14c62c21f5172dbbe444e3d0dd). The root cause for this is the `defaultProps` problem.

Hope this helps.

**Comment by eVoloshchak** [comment]
If it was caused by https://github.com/Expensify/App/commit/a453d1e610a5dc14c62c21f5172dbbe444e3d0dd and not by https://github.com/Expensify/App/pull/17310, we should fix this in a separate PR, since that's a separate issue

**Comment by akinwale** [comment]
> If it was caused by [a453d1e](https://github.com/Expensify/App/commit/a453d1e610a5dc14c62c21f5172dbbe444e3d0dd) and not by #17310, we should fix this in a separate PR, since that's a separate issue

Yeah, that's why I referred to this comment: https://github.com/Expensify/App/issues/17656#issuecomment-1515237627

**Comment by eVoloshchak** [comment]
Oh, ok, thanks for linking the comment. No problem, let's fix both issues in one PR

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.3.5-6 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/17639


If no regressions arise, payment will be issued on 2023-05-03. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@eVoloshchak] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@eVoloshchak] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@eVoloshchak] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@eVoloshchak] Determine if we should create a regression test for this bug.
- [ ] [@eVoloshchak] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@strepanier03] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by strepanier03** [comment]
I'm back from out of office starting today and am working on catching up with this now. I'll update shortly.

**Comment by strepanier03** [comment]
@eVoloshchak and @akinwale - I have hired you both in Upwork for this job. 

----

Eugene, if you can finish up the BugZero checklist, I can update the comment and do the reg test GH if needed. 

----

I'll check back in later this afternoon to pay what I can.

**Comment by akinwale** [comment]
@mountiny @eVoloshchak The [point I raised](https://github.com/Expensify/App/issues/17656#issuecomment-1516691777)  with the `applyStrikeThrough` method in `OfflineWithFeedback` still needs to be filed as a separate issue. Shall I post it in Slack?

**Comment by strepanier03** [comment]
@akinwale - I have paid you via Upwork and ended the contract, thank you again!

**Comment by eVoloshchak** [comment]
-  The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/16919

-  The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/16919#issuecomment-1536550742

-  A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: I don't think an additional discussion is needed, this is a visual bug, we already have the following item in reviewer checklist: `I turned off my network connection and tested it while offline to ensure it matches the expected behavior`

- Determine if we should create a regression test for this bug. We already have a test called [Delete Workspace (Online/Offline)](https://expensify.testrail.io/index.php?/cases/view/1971214), I think it should just be expanded with **Offline** section.  We can add a step 10 `In offline, verify that deleted workspace name is crossed out`

**Comment by strepanier03** [comment]
@mountiny, I did my normal analysis for the speed bonus and didn't think this was eligible. Below is was my review for the speed bonus:

> Speed bonus analysis: PR submitted on April 18 / PR merged on April 24 = 4 business days not eligible.

Is there anything I'm missing with this where we should make an exception and should pay the bonus?

**Comment by mountiny** [comment]
I agree not here, it was also a partially regression although hard to catch. I think we are good to go here. I am there will be many more PRs to pay for everyone, thanks fo ryou hard work!
